include/fi_util.h

```diff
@@ -295,6 +295,38 @@ struct util_event {
 int fi_eq_create(struct fid_fabric *fabric, struct fi_eq_attr *attr,
 		 struct fid_eq **eq_fid, void *context);
 
+/*
+ * Buffer Pool
+ */
+
+struct util_buf_pool {
+	size_t size;
+	size_t max_num;
+	size_t chunk_num;
+	size_t alignment;
+	size_t num_allocated;
+	size_t num_used;
+
+	struct slist buf_list;
+	struct slist region_list;
+};
+
+struct util_buf_region {
+	struct slist_entry entry;
+	char *mem_region;
+};
+
+union util_buf {
+	struct slist_entry entry;
+	uint8_t data[0];
+};
+
+struct util_buf_pool *util_buf_pool_create(size_t size, size_t alignment,
+					   size_t max_num, size_t chunk_num);
+int util_buf_pool_destroy(struct util_buf_pool *pool);
+
+void *util_buf_get(struct util_buf_pool *pool);
+void util_buf_release(struct util_buf_pool *pool, void *buf);
```

Comment:
Please move these into a separate header file (fi_bufpoll.h or fi_mem.h -- I'm leaning to fi_mem.h).